mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -264,6 +268,96 @@ class AdjointGenerator {
         }
     }
 
+    void visitOperation(func::CallOp callOp, OpBuilder &builder)
+    {
+        // Get the the original function
+        SymbolRefAttr symbol = dyn_cast_if_present<SymbolRefAttr>(callOp.getCallableForCallee());
+        func::FuncOp funcOp =
+            dyn_cast_or_null<func::FuncOp>(SymbolTable::lookupNearestSymbolFrom(callOp, symbol));
+
+        std::optional<Value> returnedQureg =
+            getQuantumReg(funcOp.front().getTerminator()->getOperands());
+        if (!returnedQureg.has_value()) {
+            // This operation is purely classical
```

Comment:
For classical function, we simply clone in the forward pass and then do nothing in the backward pass.

For quantum function (one register returned), we do nothing in the forward pass and then apply the backward pass.

For a mix of quantum and classical, we do not support it. There is an assert in the forward pass for that case.

Reply:
I have added an assert for the quantum and classical in the backward pass, just in case. Let me know what you think

Reply:
Thanks, from my side this looks fine. I'll let @erick-xanadu double check.

Reply:
Don't our quantum nodes also have a qnode attribute function? I think it is fine, but we should assign meaning to it and that meaning is likely that it is a quantum function returning a quantum argument. This is a meta comment, but in general I would prefer if we check if the function has a qnode attribute instead of looking at the return types.